Use chacha in `get_secure_random_bytes()`

danielgranhao opened the pull request:
Addresses #1958. WIP

tnull commented on lightning/src/chain/keysinterface.rs:
```diff
@@ -1236,14 +1231,16 @@ impl KeysManager {
-		sha.input(b"Unique Secure Random Bytes Salt");
-		Sha256::from_engine(sha).into_inner()
+		let mut random_bytes = [0; 32];
+		ChaCha20::new(&self.seed, &nonce).process_in_place(&mut random_bytes);
```
Rather than creating a new `ChaCha20` instance on each call, can't we keep the state in a corresponding field once and here only do the `process_in_place` part? Additionally, this would allow us to get rid of the iterating `index`.

danielgranhao replied:
Applied in [d3f14e7](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/d3f14e7e2951892a70f72a4f2234416e1a10cb1f)

tnull replied:
Thanks for the fix! While it seems to be the most straightforward,  I unfortunately suspect that taking a `Mutex` lock is likely too slow and we'd probably want to explore alternatives, e.g. the one sketched above.

As this PR aims to improve perfomance, could you add a benchmark test that would allow us to compare the approaches head-to-head (old vs. Mutex vs. iterating nonce)?

danielgranhao replied:
True. I've also discovered that the `Mutex` isn't even a real option due to the `no-std` feature. I will work on the benchmark test to compare the approaches but replace the Mutex one with a simple spin-lock.

tnull replied:
> True. I've also discovered that the `Mutex` isn't even a real option due to the `no-std` feature. I will work on the benchmark test to compare the approaches but replace the Mutex one with a simple spin-lock.

Ah, but regarding that you could just use `crate::sync::Mutex`, which chooses the right one based on `std`/`no-std`. Still, probably not the way to go.

danielgranhao replied:
Ok, I've found the time to come back to this. I've only noticed now that I didn't read quite well your first message. That's why I just went ahead with the `Mutex` based approach...

I've added the benchmark test in [e3622f1](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/e3622f136a886dfc84242d7a9872f2f6d1aded8b), followed by commit [fac7b4e](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/fac7b4e02777819868f68d539506ef70ed6ed36a) with the `Mutex` approach, finally followed by [d943b28](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/d943b2896e25b75c4b13ffcba8982da9e86ccc1e) with the iterating nonce approach. On my machine, the `Mutex` ends up being faster than the iterating nonce one by about 30-40%, but I'll wait for the CI to run to confirm my observation :)

danielgranhao replied:
I should have thought of that! Now the results make more sense :) The `Mutex` approach is 3x slower when 5 threads compete for the lock. 

Values on my machine for 100 calls:
- Single-thread `Mutex` -> ~ 6 us
- Single-thread iterating nonce -> ~10 us
- 5 competing threads `Mutex` -> ~30 us
- 5 competing threads iterating nonce -> ~90 us
- 5 competing threads original approach -> ~1900 us

It's a trade-off between getting a higher throughput implementation in high vs. low contention scenarios. I think it makes sense to go with the iterating nonce approach. Despite not reaching the highest throughput, it achieves significantly higher throughput when there's more competition.

I've pushed [3266726](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/326672612291e59b0612dcc71ff682d6cb3383cd) with the updated benchmark and [17860ac](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/17860ac0f85b82ad4df761a6ef0cd996903c1229) with the iterating nonce approach.

Thank you for the feedback and help.

With the new implementation, other tests are failing, maybe for the same reason as the one addressed in #1984. I will have a look later.

danielgranhao replied:
I'm sorry, somehow I switched the 2 values by mistake 🤦 I edited the comment now to get it fixed. I will push the 2 versions soon, I have them locally.

danielgranhao replied:
> I will push the 2 versions soon, I have them locally.

Done!

tnull commented on lightning/src/chain/keysinterface.rs:
```diff
@@ -1236,14 +1231,16 @@ impl KeysManager {
-		sha.input(&child_privkey.private_key[..]);
+		let secs_plus_index = self.starting_time_secs + index as u64;
+		let mut nonce = Vec::new();
+		nonce.append(&mut secs_plus_index.to_be_bytes().to_vec());
```
Adding an iterating index to `starting_time_secs` feels like it has some potential for generating duplicate nonces. 

Note that our docs currently state "`starting_time` isn’t strictly required to actually be a time, but it must absolutely, without a doubt, be unique to this instance. ie if you start multiple times with the same `seed`, `starting_time` must be unique to each run", which users might interpret as "I can leave the `starting_time_nanos` field 0, as long as I ensure the first part of it is unique". This unlikely scenario however might result in duplicate nonces, which should under no circumstances ever happen under the same seed. 

That said, I believe this would easily be resolved if we kept the `ChaCha20` state around, see below comment.

danielgranhao replied:
It makes complete sense; thank you for pointing it out. Fixed in [d3f14e7](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/d3f14e7e2951892a70f72a4f2234416e1a10cb1f).

tnull reviewed (commented):
Github won't let me get there, but note that `rand_bytes_unique_start` is still initialized but now unused:

https://github.com/danielgranhao/rust-lightning/blob/a512d3dda038aec465122aeae2de4844c4cacfca/lightning/src/chain/keysinterface.rs#L1020-L1023

danielgranhao commented:
@tnull Thank you for your comments. Somehow I didn't notice the unused value warnings. In [d3f14e7](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/d3f14e7e2951892a70f72a4f2234416e1a10cb1f) I also removed `rand_bytes_unique_start`. 

I'm still missing fixing the failing test `ln::functional_tests::test_duplicate_payment_hash_one_failure_one_success()`. I will try to look a bit more into it, but if anybody has any intuition about the problem, suggestions on what to look for would be highly appreciated!

jkczyz commented on lightning/src/chain/keysinterface.rs:
```diff
@@ -1015,15 +1014,14 @@ impl KeysManager {
-				rand_bytes_unique_start.input(&starting_time_secs.to_be_bytes());
-				rand_bytes_unique_start.input(&starting_time_nanos.to_be_bytes());
-				rand_bytes_unique_start.input(seed);
+				let mut nonce = Vec::new();
+				nonce.append(&mut starting_time_secs.to_be_bytes().to_vec());
+				nonce.append(&mut starting_time_nanos.to_be_bytes().to_vec());
+				let chacha = Mutex::new(ChaCha20::new(seed, &nonce));
```
I think we can avoid three vec allocations by having a fixed array and copying the bytes into it.

danielgranhao replied:
Thanks! I've changed to that approach in [fac7b4e](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/fac7b4e02777819868f68d539506ef70ed6ed36a).

danielgranhao commented on lightning/src/chain/keysinterface.rs:
```diff
@@ -979,9 +979,8 @@ pub struct KeysManager {
-	rand_bytes_child_index: AtomicUsize,
-	rand_bytes_unique_start: Sha256State,
+	rand_bytes_unique_start: [u8; 32],
+	rand_bytes_index: AtomicUsize,
```
Applied in [6472af6](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/6472af6770a83af6696d467fa9c7b5784d50eb83)

danielgranhao commented on lightning/src/chain/keysinterface.rs:
```diff
@@ -1469,3 +1461,38 @@ impl PhantomKeysManager {
+
+		for _ in 1..5 {
+			let keys_manager_clone = Arc::clone(&keys_manager);
+			thread::spawn(move || {
```
yep, ty! Applied in [34fa5af](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/34fa5afa75afc24838c351cf2d56c9e844353c2b)

danielgranhao commented on lightning/src/chain/keysinterface.rs:
```diff
@@ -1027,15 +1026,13 @@ impl KeysManager {
-				rand_bytes_unique_start.input(&starting_time_nanos.to_be_bytes());
-				rand_bytes_unique_start.input(seed);
+				let mut rand_bytes_unique_start = [0u8; 32];
+				rand_bytes_unique_start[..8].copy_from_slice(&starting_time_secs.to_be_bytes());
```
Ok, if I understood you correctly, this is what you propose -> [6472af6](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/6472af6770a83af6696d467fa9c7b5784d50eb83)

danielgranhao commented:
I've rebased to take in the changes from #1984. I've added [d08d4f8](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/d08d4f8000eaf5806c952e4071f4926d0e7ae461) just to check that fixing `test_bump_penalty_txn_on_revoked_htlcs` will fix the remaining CI checks. I suppose it needs an improvement similar to the one on `test_duplicate_payment_hash_one_failure_one_success`.

danielgranhao commented on lightning/src/chain/keysinterface.rs:
```diff
@@ -1030,10 +1031,11 @@ impl KeysManager {
-				nonce[..8].copy_from_slice(&starting_time_secs.to_be_bytes());
-				nonce[8..12].copy_from_slice(&starting_time_nanos.to_be_bytes());
-				let chacha = Mutex::new(ChaCha20::new(seed, &nonce));
+				let mut rand_bytes_unique_start = Sha256::engine();
```
Sure, applied in [1456770](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/1456770e896fabe25b4b7141f62bd3e294dd1bd4).

danielgranhao commented on lightning/src/chain/keysinterface.rs:
```diff
@@ -1027,15 +1027,15 @@ impl KeysManager {
+				let mut rand_bytes_engine = Sha256::engine();
+				rand_bytes_engine.input(&starting_time_secs.to_be_bytes());
+				rand_bytes_engine.input(&starting_time_nanos.to_be_bytes());
+				rand_bytes_engine.input(seed);
```
Sure :) Done in [f19821d](https://github.com/lightningdevkit/rust-lightning/pull/1974/commits/f19821dac4114f4329d997b4087141d8080821f3)
